posts/inside-rust/2020-12-XX-ffi-unwind-longjmp.md

```diff
@@ -0,0 +1,214 @@
+return a new `Drop` resource (making the annotated function no longer a POF),
+it may be safe, as long as we guarantee that the annotated function cannot be
+canceled while the un-annotated function is still on the stack; i.e.,
+cancelation must happen during an active call to an annotated cancelable
+function.
+
+Most importantly, we do not have a plan for how to indicate that a
+non-annotated function can safely call an annotated function. The example of
+using `setjmp` to ensure that a `longjmp` will not discard a stack frame is
+non-trivial:
+
+* `setjmp` is not a function but a C macro. There is no way to call it directly
+  in Rust.
+* `setjmp` does not prevent arbitrary `longjmp`s from crossing over a frame,
+  the way C++'s `catch` can catch any exception. Instead, `setjmp` creates an
+  object of type `jmp_buf`, which must be passed to `longjmp`; this causes the
+  jump to stop at the corresponding `setjmp` call.
+
+And, of course, `setjmp`/`longjmp` is not the only example of such a mechanism!
+Thus, there is probably no way for the compiler to guarantee that this is safe,
+and it's unclear what heuristics could be applied to make it as safe as
+possible.
+
+### Examples
+
+Let us use `#[pof-longjmp]` as a placeholder for the annotation indicating a
+function that can be safely deallocated, and let us assume that the following
+function is a wrapper around `longjmp`:
+
+```rust
+extern "C" {
+    #[pof-longjmp]
+    fn longjmp(CJmpBuf) -> !;
+}
+```
+
+The compiler would not allow this:
+
+```rust
+fn has_drop(jmp_buf: CJmpBuf) {
+    let s = "string data".to_owned();
+    unsafe { longjmp(jmp_buf); }
+    println!("{}", s);
+}
+```
+
+Here, `s` implements `Drop`, so `has_drop` is not a POF. Since `longjmp` is
+annotated `#[pof-longjmp]`, the un-annotated function `has_drop` can't call it
+(even in an `unsafe` block). If, however, `has_drop` is annotated:
+
+```rust
+#[pof-longjmp]
+fn has_drop(jmp_buf: CJmpBuf) {
+    let s = "string data".to_owned();
+    unsafe { longjmp(jmp_buf); }
+    println!("{}", s);
+}
+```
+
+...there is a different error: `#[pof-longjmp]` can only be applied to POFs,
+and since `s` implements `Drop`, `has_drop` is not a POF.
+
+An example of a permissible `longjmp` call would be:
+
+```rust
+#[pof-longjmp]
+fn no_drop(jmp_buf: CJmpBuf) {
+    let s = "string data";
+    unsafe { longjmp(jmp_buf); }
+    println!("{}", s);
+}
+```
+
+## Difference from the "never" type
+
+Above, a Rust call to C's `longjmp` function was declared with the "never"
+return type, `!`. A function with this return type cannot return normally,
+which sounds superficially similar to functions that can be "canceled".
+However, cancelation is more restrictive: `-> !` usually indicates an infinite
+loop. This cannot cause a resource leak, so it is safe for non-POFs.
+
+## Join us!
+
+If you would like to help us create this specification and write an RFC for it,
+please join us in [zulip][proj-group-zulip]!
+
+#### Footnotes
```

Comment:
Does the blog have built-in footnote support?

Reply:
I...don't remember :)